dartagnan/src/main/java/com/dat3m/dartagnan/encoding/WmmEncoder.java

```diff
@@ -645,12 +644,13 @@ public Void visitSyncFence(Relation syncFence) {
             for (int i = 0; i < allFenceSC.size() - 1; i++) {
                 Event x = allFenceSC.get(i);
                 for (Event z : allFenceSC.subList(i + 1, allFenceSC.size())) {
-                    String scope1 = Tag.PTX.getScopeTag(x);
-                    String scope2 = Tag.PTX.getScopeTag(z);
+                    String scope1 = Tag.getScopeTag(x, program.getArch());
+                    String scope2 = Tag.getScopeTag(z, program.getArch());
                     if (!scope1.equals(scope2) || scope1.isEmpty()) {
                         continue;
                     }
-                    if (!((ScopedThread) x.getThread()).sameAtHigherScope(((ScopedThread) z.getThread()),scope1)) {
+                    if (!x.getThread().hasScope() || !z.getThread().hasScope() ||
+                            !x.getThread().getScopeHierarchy().sameAtHigherScope((z.getThread().getScopeHierarchy()),scope1)) {
```

Comment:
I think you could remove from `allFenceSC` all events that are not part of a scoped thread and then get rid of the first two checks here

Reply:
Updated